content/en/blog/2024/inpod-traffic-redirection-ambient/index.md

```diff
@@ -0,0 +1,217 @@
+be included in the ambient mesh, is detected by the istio-cni node agent.
+- The istio-cni node agent then establishes network redirection rules in the top-level node network namespace, such that
+packets entering or leaving the enrolled pod would be intercepted and redirected to that node’s ztunnel on the relevant
+proxy ports (15008, 15006, or 15001).
+
+This meant that for a packet created by a mesh-enrolled pod, that packet would leave that source pod, enter the node’s
+top-level network namespace, and then be intercepted and redirected to that node’s ztunnel (running in its own network
+namespace) for proxying to the destination pod, with the return trip being similar.
+
+This model worked well enough as a placeholder for the initial Ambient launch, but as mentioned, it has a fundamental
+problem - there are many CNI implementations, and in Linux there are many fundamentally different and incompatible ways
+in which you can configure how packets get from one network namespace to another. You can use tunnels, overlay networks,
+you can go through the host network namespace, or you can bypass it. You can go through the Linux user space networking stack,
+or you can skip it and shuttle packets back and forth in the kernel space stack, etc etc. And for every possible approach,
+there’s probably a CNI implementation out there that makes use of it.
+
+Which meant that with the previous redirection approach, there were lots of CNI implementations Ambient simply wouldn’t
+work with, given its reliance on node network namespace packet redirection - any CNI that didn’t route packets thru the
+node network namespace would need a different redirection implementation. And even for CNIs that did do this, we would
+have unavoidable and potentially unresolvable problems with conflicting node-level rules - do we intercept before the CNI,
+or after? Will some CNIs break if we do one, or the other, and they aren’t expecting that? Where and when is NetworkPolicy
+enforced, since NetworkPolicy must be enforced in the node network namespace? Do we need lots of code to special-case
+every popular CNI?
+
+### Istio Ambient Traffic Redirection: The New Model
+
+In the current ambient model, this is how mesh enrollment works:
+- A Kubernetes pod (existing or newly-started), with labels indicating it should be enrolled in the ambient mesh, is detected by
+the istio-cni node agent.
+  - If a *new* pod is started that should be enrolled, a standard CNI plugin (as installed and managed by the istio-cni agent)
+is triggered by the CRI, used to push a new pod event to the node’s istio-cni agent, and block pod startup until redirection
+is established. Since CNI plugins are invoked by the CRI as early as possible in the Kubernetes pod creation process,
+this ensures that we can establish traffic redirection early enough to prevent traffic escaping during startup,
+without relying on things like init containers.
```

Comment:
```suggestion

is triggered by the CRI. This plugin is used to push a new pod event to the node’s `istio-cni` agent, and block pod startup until the agent successfully configures redirection. Since CNI plugins are invoked by the CRI as early as possible in the Kubernetes pod creation process, this ensures that we can establish traffic redirection early enough to prevent traffic escaping during startup, without relying on things like init containers.
```